src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,257 @@
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.CodeGeneration;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertLocalFunctionToMethod
+{
+    using static SyntaxFactory;
+    using static SyntaxKind;
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertLocalFunctionToMethodCodeRefactoringProvider)), Shared]
+    internal sealed class CSharpConvertLocalFunctionToMethodCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+            if (!node.IsKind(SyntaxKind.LocalFunctionStatement, out LocalFunctionStatementSyntax localFunction))
+            {
+                return;
+            }
+
+            if (localFunction.ContainsDiagnostics)
+            {
+                return;
+            }
+
+            if (!localFunction.Identifier.Span.IntersectsWith(context.Span))
+            {
+                return;
+            }
+
+            if (!localFunction.Parent.IsKind(SyntaxKind.Block, out BlockSyntax parentBlock))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c => UpdateDocumentAsync(root, document, parentBlock, localFunction, c)));
+        }
+
+        private static async Task<Document> UpdateDocumentAsync(
+            SyntaxNode root,
+            Document document,
+            BlockSyntax parentBlock,
+            LocalFunctionStatementSyntax localFunction,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declaredSymbol = semanticModel.GetDeclaredSymbol(localFunction, cancellationToken);
+
+            var dataFlow = semanticModel.AnalyzeDataFlow(localFunction.Body);
+
+            var isStatic =
+                declaredSymbol.ContainingType.IsStatic ||
+                declaredSymbol.ContainingSymbol.IsStatic ||
+                dataFlow.Captured.All(capture => !capture.IsThisParameter());
+
+            var capturedTypeParameters = GetCapturedTypeParameters(declaredSymbol);
+            var unusedTypeParameters = capturedTypeParameters.ToList();
+            foreach (var id in localFunction.DescendantNodes().OfType<IdentifierNameSyntax>())
+            {
```

Comment:
not totally convinced this is right... but not able to think of a different case.

Reply:
Yeah, when captures are generic and we didn't spell out their types anywhere..